src/Control/Middleware/HTTPCacheControlMiddleware.php

```diff
@@ -0,0 +1,776 @@
+    public static function reset()
+    {
+        Injector::inst()->unregisterNamedObject(__CLASS__);
+    }
+
+    /**
+     * @return int
+     */
+    protected function getForcingLevel()
+    {
+        if (isset($this->forcingLevel)) {
+            return $this->forcingLevel;
+        }
+        return $this->config()->get('defaultForcingLevel');
+    }
+
+    /**
+     * Generate vary http header
+     *
+     * @param HTTPResponse $response
+     * @return string|null
+     */
+    protected function generateVaryHeader(HTTPResponse $response)
+    {
+        // split the current vary header into it's parts and merge it with the config settings
+        // to create a list of unique vary values
+        $vary = $this->getVary();
+        if ($response->getHeader('Vary')) {
+            $vary = $this->combineVary($vary, $response->getHeader('Vary'));
+        }
+        if ($vary) {
+            return implode(', ', $vary);
+        }
+        return null;
+    }
+
+    /**
+     * Generate Last-Modified header
+     *
+     * @return string|null
+     */
+    protected function generateLastModifiedHeader()
+    {
+        if (!$this->modificationDate) {
+            return null;
+        }
+        return gmdate('D, d M Y H:i:s', $this->modificationDate) . ' GMT';
+    }
+
+    /**
+     * Generate Expires http header
+     *
+     * @return null|string
+     */
+    protected function generateExpiresHeader()
+    {
+        $maxAge = $this->getDirective('max-age');
+        if ($maxAge === false) {
+            return null;
+        }
+
+        // Add now to max-age to generate expires
+        $expires = DBDatetime::now()->getTimestamp() + $maxAge;
+        return gmdate('D, d M Y H:i:s', $expires) . ' GMT';
+    }
+
+    /**
+     * Update state based on current request and response objects
+     *
+     * @param HTTPRequest $request
+     * @param HTTPResponse $response
+     */
+    protected function augmentState(HTTPRequest $request, HTTPResponse $response)
+    {
+        // If sessions exist we assume that the responses should not be cached by CDNs / proxies as we are
+        // likely to be supplying information relevant to the current user only
+        if ($request->getSession()->getAll()) {
+            // Don't force in case user code chooses to opt in to public caching
+            $this->privateCache();
+        }
+
+        // IE6-IE8 have problems saving files when https and no-cache/no-store are used
+        // (http://support.microsoft.com/kb/323308)
+        // Note: this is also fixable by ticking "Do not save encrypted pages to disk" in advanced options.
+        if ($request->getScheme() === 'https' &&
+            preg_match('/.+MSIE (7|8).+/', $request->getHeader('User-Agent')) &&
+            strstr($response->getHeader('Content-Disposition'), 'attachment;') == true &&
+            ($this->hasDirective('no-cache') || $this->hasDirective('no-store'))
+        ) {
+            $this->privateCache(true);
```

Comment:
We no longer support IE< 11, so should we keep this legacy code around? I'd vote to remove it.

Keeping it means we technically vary responses based on UA, which is not in our Vary header any longer. We'd probably want to add a max-age here too(?)

Reply:
We no longer support it in the CMS, but if this affects or can be used in the frontend then we should keep it IMO

Reply:
OK, for how long? When can we safely remove IE hacks from our codebase?

Reply:
Hmm good point. I was skimming and also missed the narrow range this actually applies to (only IE 7 & 8, while running over HTTPS, plus certain headers/directives). Devs can trigger this behaviour themselves too, so I’m convinced it’s enough of an edge case to discard in core code given how old those browser are.

Reply:
Yeah, someone can add their own middleware to deal with this edgecase ;)